src/main/java/com/spartronics4915/frc2019/vision/GoalTrack.java

```diff
@@ -0,0 +1,148 @@
+package com.spartronics4915.frc2019.vision;
+
+import edu.wpi.first.wpilibj.Timer;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.TreeMap;
+
+import com.spartronics4915.lib.geometry.Translation2d;
+import com.spartronics4915.frc2019.Constants;
+
+/**
+ * A class that is used to keep track of all goals detected by the vision
+ * system. As goals are detected/not detected anymore by the vision system,
+ * function calls will be made to create, destroy, or update a goal track.
+ * 
+ * This helps in the goal ranking process that determines which goal to fire
+ * into, and helps to smooth measurements of the goal's location over time.
+ * 
+ * @see GoalTracker.java
+ */
+public class GoalTrack
+{
+
+    Map<Double, Translation2d> mObservedPositions = new TreeMap<>();
+    Translation2d mSmoothedPosition = null;
+    int mId;
+
+    private GoalTrack()
+    {
+    }
+
+    /**
+     * Makes a new track based on the timestamp and the goal's coordinates (from
+     * vision)
+     */
+    public static GoalTrack makeNewTrack(double timestamp, Translation2d first_observation, int id)
+    {
+        GoalTrack rv = new GoalTrack();
+        rv.mObservedPositions.put(timestamp, first_observation);
+        rv.mSmoothedPosition = first_observation;
+        rv.mId = id;
+        return rv;
+    }
+
+    public void emptyUpdate()
+    {
+        pruneByTime();
+    }
+
+    /**
+     * Attempts to update the track with a new observation.
+     * 
+     * @return True if the track was updated
+     */
+    public boolean tryUpdate(double timestamp, Translation2d new_observation)
+    {
+        if (!isAlive())
+        {
+            return false;
+        }
+        double distance = mSmoothedPosition.inverse().translateBy(new_observation).norm();
+        // XXX: What does kMaxTrackerDistance mean?
```

Comment:
I belive that `kMaxTrackerDistance` is the distance between the last known position of this `GoalTrack`, and the new observation.